### ISSUE TITLE
Unable to live debug multi-locale programs using --gdb

### ISSUE THREAD

**Issue by sanish81** [chapel-lang/chapel#16075]
### Summary of Problem

Attempting to launch a multi-locale program either with the "--gdb" flag or with CHPL_COMM_USE_GDB set, results in the following error `<command-line arg>:1: error: Unexpected flag:  "-e"<command-line arg>:1: error: Unexpected flag:  "-e"`

**Compile command:**

`chpl -g my_app.chpl`

**Execution command:**

`./my_app --gdb -nl 2`

### Configuration Information

<pre>`$ chpl --version
chpl version 1.20.1.f753d68
Copyright (c) 2004-2019, Cray Inc.  (See LICENSE file for more details)`</pre>
<pre>`$ printchplenv --anonymize
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: unknown
CHPL_LOCALE_MODEL: numa *
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: smp *
  CHPL_GASNET_SEGMENT: fast
CHPL_TASKS: qthreads
CHPL_LAUNCHER: smp
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: intrinsics
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_LLVM: none
CHPL_AUX_FILESYS: none`</pre>

**Comment by gbtitus** [comment]
The smp launcher differs from the others in that it doesn't invoke any system launcher to run the `_real` executable.  Instead all it has to do is to set an environment variable to the desired number of program instances and run the `_real` directly.  Unfortunately the logic common to all launchers which handles invoking a debugger assumes that there is a system launcher wrapping the `_real`, whose command line can be adjusted to wrap an `xterm` instead.  This is the root cause of the problem.  Hopefully there is an elegant way for the common launcher code to recognize that no system launcher is involved, and adjust the command line appropriately.

**Comment by bradcray** [comment]
@gbtitus: Could the common launcher code check the CHPL_LAUNCHER setting, see that it's the smp case, and special-case it?

**Comment by e-kayrakli** [comment]
I have been using the UDP conduit for multilocale debugging without any significant issue. It may be a workaround for this issue.

**Comment by gbtitus** [comment]
> Could the common launcher code check the CHPL_LAUNCHER setting ... ?

I think so.  We don't pass the `CHPL_MAKE_LAUNCHER` setting into the launcher library build as a predefined preprocessor variable currently, but we could.  Or perhaps we could add the `CHPL_LAUNCHER` setting to the runtime build's `chpl-env-gen.h` file and then `#include` that into the launcher.  Checking at execution time would be harder because the actual setting is from the chplenv python scripts rather than the environment; i.e. you could imagine an execution environment in which the chplenv script chose 'smp' as the launcher when `CHPL_LAUNCHER` wasn't set at all.  We'd have to run the python chplenv script out of the launcher to get that right.  Checking when we're compiling the launcher library definitely seems easier.

**Comment by gbtitus** [comment]
> I have been using the UDP conduit for multilocale debugging without any significant issue. It may be a workaround for this issue.

Yes, that's what they've switched to now, which somewhat relieves the pressure to get this working.

**Comment by bradcray** [comment]
> Checking at execution time would be harder

Agreed, I think this is something we'd want to decide at compile-time.  Either of the techniques you suggest for threading the information into the source seem reasonable to me.

**Comment by jabraham17** [comment]
The following patch is enough to get this working for me

```
diff --git a/runtime/src/launch/smp/launch-smp.c b/runtime/src/launch/smp/launch-smp.c
index fe626ae61ca..c8863f2b6ea 100755
--- a/runtime/src/launch/smp/launch-smp.c
+++ b/runtime/src/launch/smp/launch-smp.c
@@ -44,8 +44,9 @@ int chpl_launch(int argc, char* argv[], int32_t numLocales,
   chpl_compute_real_binary_name(argv[0]);
   snprintf(baseCommand, sizeof(baseCommand), "%s", chpl_get_real_binary_name());
 
-  return chpl_launch_using_exec(baseCommand,
-                                chpl_bundle_exec_args(argc, argv, 0, NULL),
+  char** commandToRun = chpl_bundle_exec_args(argc, argv, 0, NULL);
+  return chpl_launch_using_exec(commandToRun[0],
+                                commandToRun,
                                 NULL);
 }
 
```

However, it only results in a single xterm window, no matter how many locales there are, making it not very useful. Given that we are moving towards other/better ways of debugging multilocale Chapel, I don't think we should worry about this. Should another solution for debugging smp programs make it's way to main (it should soon), I am going to consider this resolved as "won't fix"

**Comment by jabraham17** [comment]
Closing as "won't fix", since https://github.com/chapel-lang/chapel/pull/27644 adds better alternatives for debugging with smp
